clusterLogRetriever fails w/ leak test

## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.



```
[2020-02-12T08:16:31.405Z] --- FAIL: TestT (13.57s)
[2020-02-12T08:16:31.405Z]     leaktest.go:143: Test TestT check-count 50 appears to have leaked: github.com/pingcap/tidb/executor.(*clusterLogRetriever).startRetrieving.func1.1()
[2020-02-12T08:16:31.405Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/cluster_reader.go:534 +0x424
[2020-02-12T08:16:31.405Z]         github.com/pingcap/tidb/util.WithRecovery(0xc019d6f728, 0x0)
[2020-02-12T08:16:31.405Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/misc.go:86 +0x57
[2020-02-12T08:16:31.405Z]         github.com/pingcap/tidb/executor.(*clusterLogRetriever).startRetrieving.func1(0xc01bae83c0, 0xc01bae83b0, 0xc01bc00d20, 0xc01afe60c0, 0xc00516d3fb, 0x2, 0xc00516d3fe, 0xf, 0xc00516d40e, 0xf)
[2020-02-12T08:16:31.405Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/cluster_reader.go:506 +0xeb
[2020-02-12T08:16:31.405Z]         created by github.com/pingcap/tidb/executor.(*clusterLogRetriever).startRetrieving
[2020-02-12T08:16:31.405Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/cluster_reader.go:505 +0xc8f
[2020-02-12T08:16:31.405Z] FAIL
[2020-02-12T08:16:31.405Z] FAIL	github.com/pingcap/tidb/executor	13.710s
```

2. What did you expect to see?



3. What did you see instead?



4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

https://github.com/pingcap/tidb/blob/56438a04cd225a7c7924e3cedb486a11159f6e4e/executor/cluster_reader.go#L534

--- a/executor/executor_test.go
+++ b/executor/executor_test.go
@@ -114,7 +114,7 @@ var _ = Suite(&testUpdateSuite{})
 var _ = Suite(&testPointGetSuite{})
 var _ = Suite(&testBatchPointGetSuite{})
 var _ = Suite(&testRecoverTable{})
-var _ = Suite(&testClusterReaderSuite{})
+var _ = Suite(&testMemTableReaderSuite{})
 var _ = SerialSuites(&testFlushSuite{})
 var _ = SerialSuites(&testAutoRandomSuite{&baseTestSuite{}})
 

--- a/executor/memtable_reader.go
+++ b/executor/memtable_reader.go
@@ -488,7 +488,12 @@ func (e *clusterLogRetriever) startRetrieving(ctx context.Context, sctx sessionc
 	}
 
 	// The retrieve progress may be abort
-	ctx, e.cancel = context.WithCancel(ctx)
+	exit := make(chan struct{})
+	ctx, cancel := context.WithCancel(ctx)
+	e.cancel = func() {
+		close(exit)
+		cancel()
+	}
 
 	var results []chan logStreamResult
 	for _, srv := range serversInfo {
@@ -506,7 +511,7 @@ func (e *clusterLogRetriever) startRetrieving(ctx context.Context, sctx sessionc
 			util.WithRecovery(func() {
 				defer close(ch)
 
-				// The TiDB provide diagnostics service via status address
+				// The TiDB provides diagnostics service via status address
 				remote := address
 				if serverType == "tidb" {
 					remote = statusAddr
@@ -531,10 +536,19 @@ func (e *clusterLogRetriever) startRetrieving(ctx context.Context, sctx sessionc
 						return
 					}
 					if err != nil {
-						ch <- logStreamResult{addr: address, typ: serverType, err: err}
+						select {
+						case ch <- logStreamResult{addr: address, typ: serverType, err: err}:
+						case <-exit:
+						}
 						return
 					}
-					ch <- logStreamResult{next: ch, addr: address, typ: serverType, messages: res.Messages}
+
+					result := logStreamResult{next: ch, addr: address, typ: serverType, messages: res.Messages}
+					select {
+					case ch <- result:
+					case <-exit:
+						return
+					}
 				}
 			}, nil)
 		}(ch, typ, address, statusAddr)
@@ -600,7 +614,7 @@ func (e *clusterLogRetriever) retrieve(ctx context.Context, sctx sessionctx.Cont
 		}
 	}
 
-	// All stream are draied
+	// All streams are drained
 	e.isDrained = e.heap.Len() == 0
 
 	return finalRows, nil

--- a/executor/memtable_reader.go
+++ b/executor/memtable_reader.go
@@ -415,7 +415,7 @@ func (h *logResponseHeap) Pop() interface{} {
 	return x
 }
 
-func (e *clusterLogRetriever) startRetrieving(ctx context.Context, sctx sessionctx.Context) ([]chan logStreamResult, error) {
+func (e *clusterLogRetriever) initialize(ctx context.Context, sctx sessionctx.Context) ([]chan logStreamResult, error) {
 	isFailpointTestModeSkipCheck := false
 	serversInfo, err := infoschema.GetClusterServerInfo(sctx)
 	failpoint.Inject("mockClusterLogServerInfo", func(val failpoint.Value) {
@@ -435,17 +435,6 @@ func (e *clusterLogRetriever) startRetrieving(ctx context.Context, sctx sessionc
 	instances := e.extractor.Instances
 	nodeTypes := e.extractor.NodeTypes
 	serversInfo = filterClusterServerInfo(serversInfo, nodeTypes, instances)
-
-	// gRPC options
-	opt := grpc.WithInsecure()
-	security := config.GetGlobalConfig().Security
-	if len(security.ClusterSSLCA) != 0 {
-		tlsConfig, err := security.ToTLSConfig()
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
-		opt = grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
-	}
 
 	var levels []diagnosticspb.LogLevel
 	for l := range e.extractor.LogLevels {
@@ -487,13 +476,27 @@ func (e *clusterLogRetriever) startRetrieving(ctx context.Context, sctx sessionc
 		Patterns:  patterns,
 	}
 
-	// The retrieve progress may be abort
-	exit := make(chan struct{})
-	ctx, cancel := context.WithCancel(ctx)
-	e.cancel = func() {
-		close(exit)
-		cancel()
+	return e.startRetrieving(ctx, sctx, serversInfo, req)
+}
+
+func (e *clusterLogRetriever) startRetrieving(
+	ctx context.Context,
+	sctx sessionctx.Context,
+	serversInfo []infoschema.ServerInfo,
+	req *diagnosticspb.SearchLogRequest) ([]chan logStreamResult, error) {
+	// gRPC options
+	opt := grpc.WithInsecure()
+	security := config.GetGlobalConfig().Security
+	if len(security.ClusterSSLCA) != 0 {
+		tlsConfig, err := security.ToTLSConfig()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		opt = grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
 	}
+
+	// The retrieve progress may be abort
+	ctx, e.cancel = context.WithCancel(ctx)
 
 	var results []chan logStreamResult
 	for _, srv := range serversInfo {
@@ -538,7 +541,7 @@ func (e *clusterLogRetriever) startRetrieving(ctx context.Context, sctx sessionc
 					if err != nil {
 						select {
 						case ch <- logStreamResult{addr: address, typ: serverType, err: err}:
-						case <-exit:
+						case <-ctx.Done():
 						}
 						return
 					}
@@ -546,7 +549,7 @@ func (e *clusterLogRetriever) startRetrieving(ctx context.Context, sctx sessionc
 					result := logStreamResult{next: ch, addr: address, typ: serverType, messages: res.Messages}
 					select {
 					case ch <- result:
-					case <-exit:
+					case <-ctx.Done():
 						return
 					}
 				}
@@ -564,7 +567,7 @@ func (e *clusterLogRetriever) retrieve(ctx context.Context, sctx sessionctx.Cont
 
 	if !e.retrieving {
 		e.retrieving = true
-		results, err := e.startRetrieving(ctx, sctx)
+		results, err := e.initialize(ctx, sctx)
 		if err != nil {
 			e.isDrained = true
 			return nil, err

--- a/executor/memtable_reader_test.go
+++ b/executor/memtable_reader_test.go
@@ -41,24 +41,24 @@ import (
 	"google.golang.org/grpc"
 )
 
-type testClusterReaderSuite struct {
+type testMemTableReaderSuite struct {
 	store kv.Storage
 	dom   *domain.Domain
 }
 
-func (s *testClusterReaderSuite) SetUpSuite(c *C) {
+func (s *testMemTableReaderSuite) SetUpSuite(c *C) {
 	store, dom, err := newStoreWithBootstrap()
 	c.Assert(err, IsNil)
 	s.store = store
 	s.dom = dom
 }
 
-func (s *testClusterReaderSuite) TearDownSuite(c *C) {
+func (s *testMemTableReaderSuite) TearDownSuite(c *C) {
 	s.dom.Close()
 	s.store.Close()
 }
 
-func (s *testClusterReaderSuite) TestMetricTableData(c *C) {
+func (s *testMemTableReaderSuite) TestMetricTableData(c *C) {
 	fpName := "github.com/pingcap/tidb/executor/mockMetricsPromData"
 	c.Assert(failpoint.Enable(fpName, "return"), IsNil)
 	defer func() { c.Assert(failpoint.Disable(fpName), IsNil) }()
@@ -117,7 +117,7 @@ func (s *testClusterReaderSuite) TestMetricTableData(c *C) {
 	}
 }
 
-func (s *testClusterReaderSuite) TestTiDBClusterConfig(c *C) {
+func (s *testMemTableReaderSuite) TestTiDBClusterConfig(c *C) {
 	// mock PD http server
 	router := mux.NewRouter()
 
@@ -413,12 +413,12 @@ func (s *testClusterReaderSuite) TestTiDBClusterConfig(c *C) {
 	}
 }
 
-func (s *testClusterReaderSuite) writeTmpFile(c *C, dir, filename string, lines []string) {
+func (s *testMemTableReaderSuite) writeTmpFile(c *C, dir, filename string, lines []string) {
 	err := ioutil.WriteFile(filepath.Join(dir, filename), []byte(strings.Join(lines, "\n")), os.ModePerm)
 	c.Assert(err, IsNil, Commentf("write tmp file %s failed", filename))
 }
 
-func (s *testClusterReaderSuite) TestTiDBClusterLog(c *C) {
+func (s *testMemTableReaderSuite) TestTiDBClusterLog(c *C) {
 	type testServer struct {
 		typ     string
 		server  *grpc.Server
@@ -850,7 +850,7 @@ func (s *testClusterReaderSuite) TestTiDBClusterLog(c *C) {
 	}
 }
 
-func (s *testClusterReaderSuite) TestTiDBClusterLogError(c *C) {
+func (s *testMemTableReaderSuite) TestTiDBClusterLogError(c *C) {
 	tk := testkit.NewTestKit(c, s.store)
 	fpName := "github.com/pingcap/tidb/executor/mockClusterLogServerInfo"
 	c.Assert(failpoint.Enable(fpName, `return("")`), IsNil)

